README.md
# Chuck Starter kit

[![dependencies Status](https://david-dm.org/d9767192/chuck-starterkit/master/status.svg)](https://david-dm.org/d9767192/chuck-starterkit/master)
[![devDependencies Status](https://david-dm.org/d9767192/chuck-starterkit/master/dev-status.svg)](https://david-dm.org/d9767192/chuck-starterkit/master?type=dev)

## Features

*  👏 **Hot reloading**
*  👏 **Code splitting**
*  👏 **Integration with lint testing, unit testing and coverage testing**
*  👏 **Integration with browser compatibility**
*  👏 **Less and css is supported**
*  👏 **Integration complex settings with one file**
*  👏 **Nice flexibility for extension**
*  👏 **Using modern plug-ins**

## Introduction

<div align="justify">Chuck starter kit is a environment for establishing your project quickly. There are many build-in features, like hot-reload, code splitting etc. If you want to adjust starter kit for fit your scenario, it retains many flexibility for you. </div>

## Requirements

node \^5.0.0

yarn \^0.23.0 or npm \^3.0.0

## Installation

Use git command:
```
git clone https://github.com/d9767192/chuck-starterkit.git
```

## Running the Project

After completing the installation step, you're ready to start the project!

```
npm start  # Start the development server (or `yarn start`)
```

While developing, you will probably rely mostly on npm start; however, there are additional scripts at your disposal:

|      Script (also yarn)       |                     Description                           |
|:------------------|:----------------------------------------------------------------------|
|     start         |  Start compile your project. The default serve url would be: http://localhost:3000|
|     build         |  Build your project as production (minify, hash etc). The default output root is dist|
|     build:lib     |  Build your library as production (cjs, umd format). The default location is src|
|     build:lib:cjs |  Build your library as common javascript. The default output root is lib|
|     build:lib:copy|  Copy files into lib folder, such as README.md, package.json etc.|
|     build:lib:umd |  Build your library as universal module definition. The default output root is lib/minify.|
|     test          |  Testing your codes. This command will run lint testing and unit testing at same time|
|     test:unit     |  Run unit testing alone                                               |
|   test:unit:watch | Run unit testing in watching mode (Trace your changing in testing immediately)|
|   test:lint       |  Run lint testing alone                                               |
|   test:lint:fix   | Run lint testing with automatically fixing errors                 |
|  test:lint:report | Run lint testing and output result as a report. The default path of report is reports |
|  test:coverage    | Run coverage testing and output report in reports directory       |
|  release          | Auto generate changelog|

Besides, there are many options you can attach after your command:

|    Option       |                              Description                                    |
|:----------------|:----------------------------------------------------------------------------|
|   --indir       | The entry directory of your main codes in your project. Default is `src`.   |
|   --main        | The file name of the application's entry point. Default is `main`           |
|   --outdir      | The name of the directory in which to emit compiled assets. Default is `dist`|
|   --testdir     | The name of the testing entry directory. Default is `tests`                 |
|   --reportdir   | The name of the testing output report directory. Default is `reports`       |
|   --template    | The template of html. It will be used if the configuration of html is true. Default is index.html|
|   --globalless  | The location of global less library. The default location is utils/style.less|
|   --libname     | Library name. The default is your project name|
|   --libdir      | The name of the library directory |
|   --port        | The port of the server. Default is 3000.                                    |
|   --publicpath  | The base path for all projects assets (relative to the website root). Default is '/'|
|   --sourcemaps  | Whether to generate sourcemaps. Default is false                            |
|   --externals   | A hash map of keys that the compiler should treat as external to the project. Default is {}|
|   --globals     | A hash map of variables and their values to expose globally. Default is {}  |
|   --outputlint  | Whether to output the lint report. Default is false.                        |
|   --fixlint     | Whether to fix the lint error automatically. Default is false.              |
|   --verbose     | Whether to enable verbose logging. Default is false.                        |
|   --dsiableless | Whether to disable less loader. Default is false.                           |
|   --dsiablecss  | Whether to disable css loader. Default is false.                            |
|   --dsiabletext | Whether to disable text loader. Default is false.                           |
|  --dsiableimage | Whether to disable image loader. Default is false.                          |
|   --dsiablefont | Whether to disable font loader. Default is false.                               |
|   --dsiablehtml | Whether to disable html loader. Default is false.                               |
|   --polyfill    | Assign the location of polyfill. By default, the babel-polyfill will be used.|
|   --imagesize   | The limit of image size. Default is 8192.                                   |
|   --fontsize    | The limit of font size. Default is 10000.                                   |
|   --threshold   | The setting of the coverage threshold. Default is { statements: 50, branches: 50, functions: 50, lines: 50}|
|   --config      | Webpack configuration file    |

All of default setting can be modified in project.config.js of root.

## Project structure

The project structure is designed by functionality of Redux-React. For more introduction of this structure can visit the offical website of Redux: [connect](hhttps://redux.js.org/docs/introduction/LearningResources.html). If you want to modify the folder structure, please feel free to do that. There is no constraint in project structure of our starter kit.

```
.
├── build                      # All build-related code
│   |── configs                # All configuration files
│   │   ├── jest.config.js     # The configuration of Jest
│   │   └── postcss.config.js  # You can put your settings here to do some post css prefix
│   |── lib                    # All libraries used in scripts
│   │   ├── logger.js          # A logger script for printing pretty log
│   │   └── path.js            # A path script for handling path issue in script
│   |── scripts                # All build-related scripts
│   │   ├── compile.js         # A script for handling code compiling
│   │   ├── eslint.js          # A script for handling lint checking
│   │   └── start.js           # A script for starting building server
│   |── server                 # Express application that provides webpack middleware
│   │   └── main.js            # Server application entry point
│   │── umd.webpack.config.js  # Configuration of webpack for umd version
│   └── webpack.config.js      # Configuration of webpack
├── public                     # Static public assets (not imported anywhere in source code)
├── src                        # Application source code
└── tests                      # Unit tests
```

## Testing

To add a unit test, create a `.spec.js` file anywhere inside of `./tests`. Jest and webpack will automatically find these files, and Enzyme will be available within your test without the need to import them.

## License

This project is licensed under the terms of the [MIT license](https://github.com/d9767192/chuck-starterkit/blob/master/LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v0.1.2](https://github.com/d9767192/chuck-starterkit/compare/v0.1.1...v0.1.2) - 2018-03-19
### Commits
- Add change log [`a1c1a39`](https://github.com/d9767192/chuck-starterkit/commit/a1c1a398955cab5410f5273b5638be261a0a7edb)
- Fix umd script [`657f027`](https://github.com/d9767192/chuck-starterkit/commit/657f0273c8a3906bd189ed4ff3c95610e417d061)

## [v0.1.1](https://github.com/d9767192/chuck-starterkit/compare/v0.1.0...v0.1.1) - 2018-03-19
### Commits
- Fix webpack config location [`f76e51f`](https://github.com/d9767192/chuck-starterkit/commit/f76e51f9b733ad185dbfdd0e6f167343479c4c09)
- Delete _config.yml [`727a45c`](https://github.com/d9767192/chuck-starterkit/commit/727a45c7b262c620f3034abf01b53147f244a1fa)
- Set theme jekyll-theme-cayman [`081866b`](https://github.com/d9767192/chuck-starterkit/commit/081866b05de3b7e8c44f4c7fefb115d849683844)

## v0.1.0 - 2018-03-19
### Commits
- Fix release script [`fd34815`](https://github.com/d9767192/chuck-starterkit/commit/fd34815e17dee619292038c9888a8db2d04e4f7a)
- Add starter kit environment [`cdc8ce3`](https://github.com/d9767192/chuck-starterkit/commit/cdc8ce3c527b080615314af9025d9ff204597462)
- Add environment [`bbbbf53`](https://github.com/d9767192/chuck-starterkit/commit/bbbbf53d0d784378cebf524e3b170c88d3b44315)



project.config.js
/* eslint import/no-extraneous-dependencies: off */
const { argv } = require('optimist');
const pkg = require('./package.json');

module.exports = {
  /** The environment to use when building the project */
  env: process.env.NODE_ENV || 'development',
  /** The full path to the project's root directory */
  basePath: __dirname,
  /** The name of the directory containing the application source code */
  inDir: argv.indir || 'src',
  /** The file name of the application's entry point */
  main: argv.main || 'index',
  /** The name of the directory in which to emit compiled assets */
  outDir: argv.outdir || 'dist',
  /** The name of the testing entry directory */
  testDir: argv.testdir || 'tests',
  /** The name of the testing output report directory  */
  reportDir: argv.reportdir || 'reports',
  /** The template of html. It will be used if the configuration of html is true */
  htmlTemplate: argv.template || 'index.html',
  /** Global less file */
  globalLess: argv.globalless || 'utils/style.less',
  /** Library name */
  libName: argv.libname || pkg.name,
  /** The name of the library directory */
  libDir: argv.libdir || 'lib',
  /** The port of the server */
  port: argv.port || 3000,
  /** The base path for all projects assets (relative to the website root) */
  publicPath: argv.publicpath || '',
  /** The main file for publish */
  publicMain: argv.publicmain || './index.js',
  /** Whether to generate sourcemaps */
  sourcemaps: argv.sourcemaps || false,
  /** A hash map of keys that the compiler should treat as external to the project */
  externals: argv.externals || {},
  /** A hash map of variables and their values to expose globally */
  globals: argv.globals || {},
  /** Whether to output the lint report */
  outputLint: argv.outputlint || false,
  /** Whether to fix the lint error automatically */
  fixLint: argv.fixlint || false,
  /** Whether to enable verbose logging */
  verbose: argv.verbose || false,
  /** Whether to use less loader */
  disableLess: argv.dsiableless || false,
  /** Whether to use css loader */
  disableCss: argv.dsiablecss || false,
  /** Whether to use text loader */
  disableText: argv.dsiabletext || false,
  /** Whether to use image loader */
  disableImage: argv.dsiableimage || false,
  /** Whether to use font loader */
  disableFont: argv.dsiablefont || false,
  /** Whether to use html loader */
  disableHtml: argv.dsiablehtml || false,
  /** Assign the location of polyfill. By default, the babel-polyfill will be used. */
  polyfill: argv.polyfill || 'babel-polyfill',
  /** The limit of image size */
  imageSize: argv.imagesize || 8192,
  /** The limit of font size */
  fontSize: argv.fontsize || 10000,
  /** The setting of the coverage threshold */
  coverageThreshold: argv.threshold || {
    statements: 50,
    branches: 50,
    functions: 50,
    lines: 50,
  },
  /** The list of modules to bundle separately from the core application code */
  vendors: [
    'react',
    'react-dom',
    'prop-types',
  ],
  /** Files to be copied into publish folder */
  copyFiles: [
    'README.md',
    'CHANGELOG.md',
    'LICENSE',
  ],
  /** Folders to be copied into publish folder */
  copyFolders: [
    'assets',
  ],
  /** webpack configuration file */
  webpack: argv.config || 'webpack.config.js',
};
